File: homy/integrations/aniworld_downloader/widgets/status/status.py
from flask import jsonify, request

from homy.cache import cache_key, should_bypass_cache, widget_cache
from homy.integration_widget_util import fetch_integration_for_widget as fetch_for_widget

WIDGET = {
    'type': 'aniworld_status',
    'name': 'AniWorld · Live Status',
    'integration_types': ['aniworld_downloader'],
    'default_size_x': 8,
    'default_size_y': 5,
    'icon': 'activity',
    'config_schema': {
        'integration_id': {'type': 'text', 'label': 'Integration ID', 'default': ''},
        'endpoint': {'type': 'text', 'label': 'Endpoint key', 'default': 'v1_status'},
    },
}

CACHE_NS = 'aniworld_status'
CACHE_TTL = 15


def register(app):
    @app.route('/api/aniworld_status/data', methods=['GET'], endpoint='widget_aniworld_status_data')
    def aniworld_status_data():
        widget_id = request.args.get('widget_id')
        if not widget_id:
            return jsonify({'configured': False, 'message': 'widget_id required'}), 400

        ck = cache_key(CACHE_NS, widget_id)
        if not should_bypass_cache():
            cached = widget_cache.get(ck)
            if cached is not None:
                return jsonify(cached)

        payload, err, code = fetch_for_widget(widget_id, 'aniworld_downloader', 'v1_status')
        if err:
            return jsonify({'configured': False, 'message': err}), code

        queue = payload.get('queue') or {}
        running = payload.get('currently_running')
        result = {
            'configured': True,
            'online': True,
            'version': payload.get('version', ''),
            'paused': bool(payload.get('paused')),
            'queue': {
                'total': queue.get('total', 0),
                'queued': queue.get('queued', 0),
                'running': queue.get('running', 0),
                'completed': queue.get('completed', 0),
                'failed': queue.get('failed', 0),
            },
            'currently_running': running,
            'progress_percent': 0,
        }
        if running:
            result['progress_percent'] = (
                running.get('overall_progress_percent')
                or (running.get('episode_progress') or {}).get('percent')
                or 0
            )

        widget_cache.set(ck, result, ttl=CACHE_TTL)
        return jsonify(result)

File: homy/integrations/aniworld_downloader/widgets/status/status.js
window.WidgetRegistry.register('aniworld_status', {
    _t(key, fallback) {
        if (!window.i18n) return fallback;
        const value = window.i18n.translate(key);
        return value === key ? fallback : value;
    },

    async render(container, widgetData) {
        container.innerHTML = '<div class="widget-loading"><div class="spinner"></div></div>';
        try {
            const data = await API.request(`/api/aniworld_status/data?widget_id=${widgetData.id}`);
            container.innerHTML = '';
            if (!data.configured) {
                container.innerHTML = `<div class="muted-text" style="padding:16px;text-align:center;">${data.message}</div>`;
                return;
            }

            const q = data.queue || {};
            const run = data.currently_running;
            const paused = data.paused
                ? this._t('aniworld_paused', 'Pausiert')
                : this._t('aniworld_active', 'Aktiv');
            const wrap = document.createElement('div');
            wrap.className = 'aniworld-status-widget';

            wrap.innerHTML = `
                <div class="aniworld-status-head">
                    <span class="aniworld-badge ${data.paused ? 'paused' : 'active'}">${paused}</span>
                    <span class="muted-text">v${data.version || '?'}</span>
                </div>
                <div class="aniworld-queue-grid">
                    <div class="aniworld-stat"><span class="num">${q.queued ?? 0}</span><span class="lbl">${this._t('aniworld_queued', 'Warteschlange')}</span></div>
                    <div class="aniworld-stat"><span class="num">${q.running ?? 0}</span><span class="lbl">${this._t('aniworld_running', 'Läuft')}</span></div>
                    <div class="aniworld-stat"><span class="num">${q.completed ?? 0}</span><span class="lbl">${this._t('aniworld_completed', 'Abgeschlossen')}</span></div>
                    <div class="aniworld-stat"><span class="num">${q.failed ?? 0}</span><span class="lbl">${this._t('aniworld_failed', 'Fehlgeschlagen')}</span></div>
                </div>
            `;

            if (run && run.title) {
                const block = document.createElement('div');
                block.className = 'aniworld-current';
                const pct = Math.min(100, parseInt(data.progress_percent, 10) || 0);
                block.innerHTML = `
                    <div class="aniworld-current-title">${run.title}</div>
                    <div class="muted-text" style="font-size:0.75rem;">
                        ${this._t('aniworld_episode', 'Episode')} ${run.current_episode || '?'}/${run.total_episodes || '?'}
                    </div>
                    <div class="aniworld-progress"><div class="fill" style="width:${pct}%"></div></div>
                `;
                wrap.appendChild(block);
            }

            container.appendChild(wrap);
        } catch (err) {
            container.innerHTML = `<div class="widget-error">${err.message}</div>`;
        }
    },
});
